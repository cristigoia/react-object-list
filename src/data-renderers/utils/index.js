export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
