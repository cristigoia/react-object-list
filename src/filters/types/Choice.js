import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

/**
 * Filter input for choice values
 */
class Choice extends React.Component {
  static propTypes = {
    /** If true, allow selection of multiple values */
    multi: PropTypes.bool,
    /** Function to be called when value changed */
    onChange: PropTypes.func.isRequired,
    /** Currently selected filter value */
    value: PropTypes.any,
    /** Placeholder text to display in input */
    placeholder: PropTypes.string,
    /** Options to choose from */
    options: PropTypes.arrayOf(PropTypes.object),
    /** Key to be used to identify label in options object */
    labelKey: PropTypes.string,
    /** Key to be used to identify value in options object */
    valueKey: PropTypes.string,
    /** Renderer to be used to display the options */
    optionRenderer: PropTypes.func,
    /** If true, data must be fetched */
    remote: PropTypes.bool,
    /** If remote true, this is called to load options */
    loadOptions: PropTypes.func,
  }

  static defaultProps = {
    multi: false,
    remote: false,
    value: null,
    placeholder: 'Any value',
    options: [],
    valueKey: 'value',
    labelKey: 'label',
    optionRenderer: null,
  }

  /**
   * Handles value change and checks selected option(s)
   * type is correct before calling on change with the values
   * @param  {object||object[]} newValue Just selected options
   */
  handleChange = newValue => {
    const {onChange, valueKey, multi} = this.props
    if (!multi) {
      if (Array.isArray(newValue)) {
        newValue = newValue[0]
      }
      onChange(newValue[valueKey])
    } else {
      if (!Array.isArray(newValue)) {
        newValue = [newValue]
      }
      onChange(newValue.map(x => x[valueKey]))
    }
  }

  render() {
    const {
      options, value, multi, placeholder,
      valueKey, labelKey, optionRenderer,
      remote, loadOptions,
    } = this.props
    const SelectComponent = remote ? Select.Async : Select
    return (
      <SelectComponent
        multi={multi}
        options={options}
        loadOptions={loadOptions}
        value={value}
        onChange={this.handleChange}
        placeholder={placeholder}
        valueKey={valueKey}
        labelKey={labelKey}
        optionRenderer={optionRenderer}
        className="objectlist-current-filter__active-status"
      />
    )
  }
}
export default Choice
