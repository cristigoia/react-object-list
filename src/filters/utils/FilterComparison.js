import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
