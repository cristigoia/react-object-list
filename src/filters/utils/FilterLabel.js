import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
