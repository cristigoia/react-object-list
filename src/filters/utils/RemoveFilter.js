import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render button used to remove a filter
 */
class RemoveFilter extends React.Component {
  static propTypes = {
    /** Function to be called to remove filter */
    onClick: PropTypes.func.isRequired,
  }

  render() {
    return (
      <button className="objectlist-button objectlist-button--delete objectlist-button--icon objectlist-button--borderless" onClick={this.props.onClick} title="Remove Filter" />
    )
  }
}

export default RemoveFilter
