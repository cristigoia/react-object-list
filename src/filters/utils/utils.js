const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
