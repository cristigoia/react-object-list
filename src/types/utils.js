function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
