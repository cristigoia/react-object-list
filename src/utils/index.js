export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'
