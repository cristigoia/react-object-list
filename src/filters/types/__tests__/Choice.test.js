import React from 'react'
import { shallow } from 'enzyme'
import { snapshotTest } from 'utils/tests'
import Choice from '../Choice'

jest.mock('react-select')

describe('Choice', () => {
  const baseProps = {
    onChange: jest.fn(),
    value: 44,
  }
  describe('Snapshots', () => {
    it('renders default', () => {
      snapshotTest(<Choice {...baseProps} />)
    })
    it('renders with options', () => {
      snapshotTest(<Choice {...baseProps} options={[{value: 1, label: 'One'}]} />)
    })
    it('renders with multiple choice', () => {
      snapshotTest(<Choice {...baseProps} multi />)
    })
    it('renders with remote options', () => {
      snapshotTest(<Choice {...baseProps} remote loadOptions={jest.fn()} />)
    })
    it('renders with custom value key', () => {
      snapshotTest(<Choice {...baseProps} valueKey="bob" />)
    })
    it('renders with custom label key', () => {
      snapshotTest(<Choice {...baseProps} labelKey="bob" />)
    })
    it('renders with custom option renderer', () => {
      const CustomRenderer = ({label, value}) => <h5>HI {label} {value}</h5>
      snapshotTest(<Choice {...baseProps} labelKey="bob" optionRenderer={CustomRenderer} />)
    })
    it('renders with placeholder', () => {
      snapshotTest(<Choice {...baseProps} placeholder="Testing testing" />)
    })
  })
  describe('Functions', () => {
    describe('handles value changing', () => {
      beforeEach(() => {
        spyOn(baseProps, 'onChange')
      })
      describe('handles single choice', () => {
        it('receives single value', () => {
          const instance = shallow(<Choice {...baseProps} />).instance()
          instance.handleChange({value: 5})
          expect(baseProps.onChange).toHaveBeenCalledWith(5)
        })
        it('receives array value', () => {
          const instance = shallow(<Choice {...baseProps} />).instance()
          instance.handleChange([{value: 9}, {value: 5}])
          expect(baseProps.onChange).toHaveBeenCalledWith(9)
        })
        it('has custom value key', () => {
          const instance = shallow(<Choice {...baseProps} valueKey="bob" />).instance()
          instance.handleChange({bob: 67})
          expect(baseProps.onChange).toHaveBeenCalledWith(67)
        })
      })
      describe('handles multiple choice', () => {
        it('receives array value', () => {
          const instance = shallow(<Choice {...baseProps} multi />).instance()
          instance.handleChange([{value: 9}, {value: 5}])
          expect(baseProps.onChange).toHaveBeenCalledWith([9, 5])
        })
        it('receives single value', () => {
          const instance = shallow(<Choice {...baseProps} multi />).instance()
          instance.handleChange({value: 5})
          expect(baseProps.onChange).toHaveBeenCalledWith([5])
        })
        it('has custom value key', () => {
          const instance = shallow(<Choice {...baseProps} multi valueKey="bob" />).instance()
          instance.handleChange([{bob: 67}])
          expect(baseProps.onChange).toHaveBeenCalledWith([67])
        })
      })
    })
  })
})
