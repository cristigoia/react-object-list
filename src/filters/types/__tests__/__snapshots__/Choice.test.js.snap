// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice Snapshots renders default 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with custom label key 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="bob"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with custom option renderer 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="bob"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={[Function]}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with custom value key 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="bob"
/>
`;

exports[`Choice Snapshots renders with multiple choice 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={undefined}
  multi={true}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with options 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={
    Array [
      Object {
        "label": "One",
        "value": 1,
      },
    ]
  }
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with placeholder 1`] = `
<Select
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={undefined}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Testing testing"
  value={44}
  valueKey="value"
/>
`;

exports[`Choice Snapshots renders with remote options 1`] = `
<Select.Async
  className="objectlist-current-filter__active-status"
  labelKey="label"
  loadOptions={[MockFunction]}
  multi={false}
  onChange={[Function]}
  optionRenderer={null}
  options={Array []}
  placeholder="Any value"
  value={44}
  valueKey="value"
/>
`;
